Move deploys csv export to an asynchronous job

rikeoka opened the pull request:
* The deploys report was failing to generate realtime with a large number of records.  The report is now generated in a background job.  The stored files is cleaned up the next time a CSV Job is run, when it is older than the lessor of 12 hours after download or 1 month after creation.

* UI changes.
Header - Added reports to Admin menu (Admin menu now shows for all users but only shows Admin options if admin, reporting is one such non-admin task):
<img width="328" alt="screen shot 2016-03-11 at 1 38 20 pm" src="https://cloud.githubusercontent.com/assets/11447948/13716249/c78f73e6-e78e-11e5-8ff8-e7614bc03ab6.png">

Recent Deploys:
![screen shot 2016-02-24 at 11 50 24 am](https://cloud.githubusercontent.com/assets/11447948/13298833/460ab8e2-daed-11e5-8592-57e2728676f6.png)

New CSV pages:
/csv
![screen shot 2016-02-24 at 11 50 59 am](https://cloud.githubusercontent.com/assets/11447948/13298853/5bfbcdda-daed-11e5-9a30-4d5030b59cf8.png)

/csv/new
![screen shot 2016-02-24 at 11 51 05 am](https://cloud.githubusercontent.com/assets/11447948/13298869/6b88a62e-daed-11e5-8878-008b3364d33c.png)

/csv/status/:id
![screen shot 2016-03-18 at 10 44 11 am](https://cloud.githubusercontent.com/assets/11447948/13886886/63e1ec58-ecf6-11e5-8cd6-a0bcb252cf00.png)
![screen shot 2016-03-18 at 10 44 48 am](https://cloud.githubusercontent.com/assets/11447948/13886908/7a427ba2-ecf6-11e5-8acb-71e0a51bfa03.png)


/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
Moved code branch from forked repo to zendesk repo to ease review per @grosser .  
Original PR https://github.com/zendesk/samson/pull/747 
JIRA: TOOLS-550

### Risks
- Level: Med

rikeoka commented:
Outstanding comment from @grosser is ".error" on file "app/jobs/csv_export_job.rb"

grosser commented on Gemfile.lock:
```diff
@@ -403,6 +403,8 @@ GEM
     sqlite3 (1.3.11)
     sse-rails-engine (1.4.0)
       rails (~> 4.0)
+    sucker_punch (2.0.1)
```
we have a restart protection in the initializers ... they should wait for sucker-push to be done too instead of killing them ?

rikeoka replied:
These jobs aren't critical so could be freely killed.  Worse case is if a server reboot happens, the user can request another export to run.  After it ages for a while, the next time a report is generated, it would be cleaned up.  Alternatively, if there is a place where something is run upon start, the exports with a pending or started status could be queued again.

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+  
+  def show
+    respond_to do |format|
+      @csv_export = CsvExport.find_by_id(params[:id])
```
we can use an actual `find` and then do `rescue_from ActiveRecord::RecordNotFound` and render custom json there

rikeoka replied:
Added to next commit.

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  def create
+    csv_export = CsvExport.create(user: current_user, content: params[:content])
```
since it is not dealing with errors a `create!` would show us the errors if something goes wrong

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+
+  def download(params)
+    csv_export = CsvExport.find_by_id(params[:id])
+    if !csv_export.nil?
```
use .find here too and rely on the `rescue_from` to handle the 404

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  def check_file(csv_export)
+    if (csv_export.ready?)
```
no need for parentheses

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+  def check_file(csv_export)
+    if (csv_export.ready?)
+      filename = full_filename(csv_export)
+      csv_export.deleted! if !File.exist?(filename)
```
if the file does not exist, we should still mark it as deleted ?

rikeoka replied:
Yes, this way the user can see the message that the file somehow got deleted, so they know to generate a new report.

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+    render json: {status: "not-found"}.to_json, :status => 404
+  end
+
+  def not_found
```
use real 404 pages ... much easier to debug

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,70 @@
+    redirect_to({action: 'index'}, flash: { error: "The CSV export does not exist." })
+  end
+
+  def full_filename(csv_export)
```
th export should know where it is stored ... not the controller ?

rikeoka replied:
moved to the model, and refactored.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,85 @@
+
+    export_completed_notify(csv_export)
+
+    rescue
```
rescue should be outdented

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,85 @@
+  
+  def export_completed_notify(csv_export)
+    if csv_export.started?
+      CsvMailer.created_email(csv_export)
```
maybe something like `notify_of_creation`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,85 @@
+  end
+  
+  def deploy_csv_export(filename)
+    @deploys = Deploy.joins(:stage).all
```
`includes` (cheaper) not `joins` ... try having nothing ... goldiloader should handle that ...

rikeoka replied:
In the original deploy export there was an issue that was causing the export to fail with deleted projects.  This was a quick fix to get it to work.
deploy 1, project a
deploy 2, project b (deleted) -> errors out

This was a simple way to make it skip over where projects were deleted and have a correct record count to see if the report failed midway in creation.

This issue shows itself in in the deploy views as well when it encounters a deploy record with a deleted project/stage.  For us deleted projects are so far back like page 5000+ that noone would run into it unless crawling.

grosser replied:
sounds good ... maybe some inline docs ... and reuse the scope instead of building it twice

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,85 @@
+  def deploy_csv_export(filename)
+    @deploys = Deploy.joins(:stage).all
+    CSV.open(filename, 'w+') do |csv|
+      csv << ["Deploy Number", "Project Name", "Deploy Sumary", "Deploy Updated", "Deploy Created", "Deployer Name", "Buddy Name", "Production Flag", Deploy.joins(:stage).count.to_s + " Deploys"]
```
very long ... break into multiple lines 
number of deploys does not really make sense as a column ... it is the size of the csv anyway ?

rikeoka replied:
It's to report number of records that should be in the report.  I couldn't find a better place for the information.  I guess it could be added to the very bottom of the report once all the records have been exported.  

The auditors care about how much rows match the condition in the system equals the number of rows in the report.

grosser replied:
we add it so they don't have to count the rows ?
... end of file sounds ok ... if we can put it in the filename, the data would be clean ...

rikeoka replied:
Filename includes when the job was started, so it would be cleaner to put at end of file.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,85 @@
+  end
+  
+  def create_export_folder
+    Dir.mkdir("#{Rails.root}/export") unless File.exist?("#{Rails.root}/export")
```
no need to check exist, `mkdir` takes care of that

rikeoka replied:
When the exist check goes away, it results in errors for the tests.  I originally thought the same where if the directory exists it would just return without any exceptions.

grosser commented on app/mailers/csv_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+    subject = "Samson CSV Export Completed"
+    url = csv_url( id: csv_export.id, format: 'csv')
+    body = "Download your CSV file at #{url}"
+    mail(to: address, subject: subject, body: body) unless (address.nil? or address.empty?)
```
handle the `email.blank?` case outside, I think otherwise it creates a strange empty email / no need to create a job

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def email
+    if user.nil?
```
`user.try :email`

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+  
+  def filename
+    filename = "#{self.content}_#{formatted_create_at}.csv"
```
no need to assign a local variable

grosser replied:
no need for self.

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,79 @@
+    filename = "#{self.content}_#{formatted_create_at}.csv"
+  end
+  
+  def pending!
```
how about only 1 method for status query and 1 for status setting ... 

```
def status?(x)
  x == :ready ? [:downloaded, :finished].include?(state) : state == x
end
```

also validate that only know states are being used

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,79 @@
+  private
+
+  def formatted_create_at
+    created_at.to_datetime.strftime "%Y%m%d_%H%M%S"
```
`created_at` is a datetime ...

grosser commented on app/views/csvs/index.html.erb:
```diff
@@ -0,0 +1,37 @@
+  <div class="deploy-csv">
+    <%= link_to "Start a CSV export job", new_csv_path %>
+  </div>
+  <% if @csv_exports.nil? %>
```
will never be nil -> `@csv_exports.none?`

grosser commented on app/views/csvs/index.html.erb:
```diff
@@ -0,0 +1,37 @@
+            <td><%= csv_export.content %></td>
+            <td><%= csv_export.created_at %></td>
+            <td><%= csv_export.status %></td>
+            <% if csv_export.finished? or csv_export.downloaded? %>
```
use `ready`

grosser replied:
prefer `||` over `or`

grosser commented on app/views/csvs/index.html.erb:
```diff
@@ -0,0 +1,37 @@
+            <td><%= csv_export.created_at %></td>
+            <td><%= csv_export.status %></td>
+            <% if csv_export.finished? or csv_export.downloaded? %>
+              <td><%= link_to "Download", csv_path(id: csv_export.id, format: 'csv') %></td>
```
`csv_path(csv_export, format: :csv)`

grosser commented on test/controllers/admin/users_controller_test.rb:
```diff
@@ -50,7 +50,7 @@
     end
     
-    describe 'a csv GET to #index' do
+    describe 'a csvs GET to #index' do
```
csv is refering to the format -> singular

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+        it "renders the page with user link & admin menu" do
+          get :new
+          assert_template :new
+          assert @response.body.include? "Create Deploys CSV Report"
```
use `@response.body.must_include` / `wont_include` to get nice error message when it fails instead of `assertion failed`

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+    end
+
+    as_a_deployer do
+      describe "a GET to new" do
```
might e easier to maintain/refactor when mixed into the normal `describe new` blocks ...

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+          assert @response.body.include? "Create Deploys CSV Report"
+          refute @response.body.include? "Download Users CSV Report"
+          assert @response.body.include? "Admin"
+          refute @response.body.include? "Environment variables"
```
why are we testing for `Environment variables` ?

rikeoka replied:
Checking that a keyword in the admin menu options (upper set) only displays for admins, since reporting of deploys may be done by someone that isn't an admin.

If there is a different thought on the permissions for reporting, it could be limited to only admins.  Ideally a permission level other than admin should be able to generate both the users and deploys reports so people that need those reports don't have to be admins.  After this only the user report will require admin, but that could probably be changed.

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+        it "renders the page without user link & limited admin menu" do
+          get :new
+          assert_template :new
+          assert @response.body.include? "Create Deploys CSV Report"
```
extract the labels to `let` blocks so they do not drift / only 1 gets changed when a test starts breaking

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+  end
+
+  as_a_admin do
+    describe "a GET to index for a user with csv exports" do
```
`for a user with csv exports` ... where is the `with no csv exports` test ?

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+        let(:format) { :json }
+        
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
```
not testing that there are any exports ...

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+  end
+
+  as_a_viewer do
+    describe "a GET to index for a user with no csv exports" do
```
this is changing 2 different things ... ideally a test should only change 1 thing
also there is no special `0 exports` logic ... so this amount of tests would be good enough ...

```Ruby
as_a_viewer do
  describe "#index" do
    it "renders html" do
      get :index, format: :html
      assert_template :index
    end

    it "renders csv" do
      get :index, format: :csv
     @response.content_type.must_equal  "application/json"
    end
  end
end
```

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+      end
+    end
+
+    describe "a GET to show for pending" do
```
please group show under 1 describe block so things don't get mixed up

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+      describe "as JSON" do
+        let(:format) { :json }
+        
+        it "renders json" do
```
can be reused as `def self.it_renders_json`

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+
+    describe "a GET to show for downloaded" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
```
`let(:csv_export)` and reuse that ... then the before can be `before { csv_export.downloaded! }`

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+      end
+      
+      describe "as deleted" do
+        it "redirects to show" do
```
extract a method ... don't copy-paste

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+      teardown do
+        @csv_exports = CsvExport.where(user_id: deployer.id)
+        @csv_exports.each do |csv_file|
+          filename = "#{Rails.root}/export/#{csv_file.id}"
```
how about: "on teardown delete all exports" ?

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+
+    describe "a Get to show as csv with file" do
+      setup do
+        post :create, content: "deploys"
```
seems a bit strange to call create in a show test ...

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+        end
+      end
+    
+      it "should create a new csv_export and redirects to status" do
```
use active verbs instead of should `it creates ...`

grosser commented on test/controllers/csvs_controller_test.rb:
```diff
@@ -0,0 +1,335 @@
+      end
+    end
+
+    describe "a Get to show as csv with file" do
```
might be simpler to consolidate all the show tests into 1 block ... then it's easier to see what is going on and usually works is being done on 1 method so only that methods block needs to be inspected and not all of the test file

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,70 @@
+  
+  describe "Job executes for deploy csv" do
+    teardown do
+      job = CsvExport.find(deploy_export_job.id)
```
same cleanup-all-exports logic ? -> make a helper and reuse

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,70 @@
+      job = CsvExport.find(deploy_export_job.id)
+      filename = "#{Rails.root}/export/#{job.id}"
+      assert job.finished?, "Not Finished"
+      assert File.exist?(filename), "File Not exist"
```
"File missing"

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,70 @@
+    it "finishes with file" do
+      CsvExportJob.perform_now(deploy_export_job.id)
+      job = CsvExport.find(deploy_export_job.id)
+      filename = "#{Rails.root}/export/#{job.id}"
```
repeated multiple times -> `job.file_path` or so

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,70 @@
+  
+  describe "Job executes for invalid csv export type" do
+    teardown do
+      job = CsvExport.find(invalid_export_job.id)
```
copy-pasted again ... `let(:job) { ... invalid job ... }`

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,70 @@
+
+    test "fails with no file" do
+      CsvExportJob.perform_now(invalid_export_job.id)
+      job = CsvExport.find(invalid_export_job.id)
```
`.reload` ?

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,60 @@
+describe CsvExport do
+  let(:user) { users(:deployer) }
+
+  describe "new CsvExport" do
```
`describe "#pending!"` ... the exporter being new does not really make these tests any different

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,60 @@
+      assert_not_empty @csv_export.filename
+    end
+
+    describe "returns users email" do
```
`describe "#email"`

rikeoka commented:
Updated UI to add filtering to the Deploys Report

<img width="594" alt="screen shot 2016-03-21 at 7 21 57 pm" src="https://cloud.githubusercontent.com/assets/11447948/13940410/65aba492-ef9a-11e5-8ac5-4f9a9a41e419.png">

I'm not too familiar with the CSS used in Samson, so if there is any existing css that could be leveraged, I'd rather use those than creating additional CSS.

grosser commented:
exporting running/cancelling does not make much sense ?
errored and failed are pretty much the same ... and cancelled is not really useful ...
how about radio button: 'all, succeeded, failed'

grosser commented:
"Configure your report with these parameters" is also redundant :)

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+      @csv_export = CsvExport.find(params[:id])
+      format.html
+      format.json { render json: @csv_export }
+      format.csv { download(params) }
```
this is finding the export once outside and once inside the params, can the export be passed to the download method ? ... or can download use @ecs_export ?

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+
+  def index
+    respond_to do |format|
+      @csv_exports = CsvExport.where(user_id: current_user.id)
```
style nitpick: assignments should stay out of the `respond_to` block ...

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def create
+    csv_export = CsvExport.create!(user: current_user, content: params[:content], filters: params.to_json)
```
storing all of params includes actions/contoller etc unnecessary things, can we store `params[:filter]` instead ?

rikeoka replied:
Moved the parser to the controller, so it can error out to the user, and only save the filter.

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    csv_export = CsvExport.create!(user: current_user, content: params[:content], filters: params.to_json)
+    csv_export.pending!
+    CsvExportJob.perform_later(csv_export.id)
+    redirect_to action: "show", id: csv_export.id
```
`redirect_to csv_export` should work

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    end
+  end
+
+  rescue_from ActiveRecord::RecordNotFound do |exception|
```
should be at the top of the file

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    respond_to do |format|
+      format.json { render json: {status: "not found"}.to_json, status: :not_found }
+      format.csv { render body: "not found", status: :not_found }
+      format.all { render file: "#{Rails.public_path}/404.html", status: :not_found }
```
`.html` instead of .all ?

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def check_file(csv_export)
+    if csv_export.status? :ready
```
same condition as above ... this would mark it at deleted and then try to download ... missing testcase ?

grosser commented on app/controllers/deploys_controller.rb:
```diff
@@ -43,8 +41,7 @@ def recent
       format.csv do
-        datetime = Time.now.strftime "%Y%m%d_%H%M"
-        send_data Deploy.to_csv, type: :csv, filename: "Deploys_#{datetime}.csv"
+        redirect_to controller: 'csvs', action: 'new'
```
can this be `new_csvs_path`  ?

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+    timeout_date = Time.now - Rails.application.config.samson.export_job.max_age
+    @csv_exports = CsvExport.where("(status = 'downloaded' AND updated_at <= :end_date) OR updated_at <= :timeout_date",
+      {end_date: end_date, timeout_date: timeout_date})
+    if @csv_exports
```
this will always be true

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+    if @csv_exports
+      @csv_exports.each do |csv_downloaded|
+        filename = csv_downloaded.full_filename
+        File.delete(filename) unless !File.exist?(filename)
```
`if File.exist?`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+      @csv_exports.each do |csv_downloaded|
+        filename = csv_downloaded.full_filename
+        File.delete(filename) unless !File.exist?(filename)
+        csv_downloaded.delete
```
call destroy instead of delete ... then the after_destroy callback of csv_export can handle file cleanup

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+    @csv_exports = CsvExport.where("(status = 'downloaded' AND updated_at <= :end_date) OR updated_at <= :timeout_date",
+      {end_date: end_date, timeout_date: timeout_date})
+    if @csv_exports
+      @csv_exports.each do |csv_downloaded|
```
might be more understandable if the local variable was `csv_export`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+  end
+  
+  def export_task(csv_export)
+    if csv_export.content == "deploys"
```
this `deploys` seems to be some magic word ... not sure where it gets set ... might be nice to use a constant like `CsvExport::FINISHED_EXPORT_MARKER`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+
+    notify_of_creation(csv_export)
+  rescue
+    #delete_file(csv_export.full_filename)
```
can go away ? ... needs a test ?

rikeoka replied:
Should the file stay on failures, so an admin can see where the report failed?  The cleanup routine will delete it later, but in some ways not having the files persist on the system is better.

grosser replied:
I'd expect errors to be in ruby land and not halfway through generating the file and we don't want to ship a half-built file ... so better get rid of it ...

rikeoka replied:
Nevermind, I can remove this.  I can create a new method in the controller for different report types if realtime.  The new is only for reports that have to be a job.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+  end
+  
+  def notify_of_creation(csv_export)
+    if csv_export.status? :started
```
can move into the `if started` block above ?

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+  
+  def notify_of_creation(csv_export)
+    if csv_export.status? :started
+      CsvMailer.created_email(csv_export).deliver_now unless (csv_export.email.nil? || csv_export.email.empty?)
```
`if csv_export.email.present?`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+
+    @deploys = Deploy.joins(:stage, :job).where(filter)
+    summary = [ "-", "Deploys", @deploys.count.to_s ]
+    filters_applied = [ "-", "Filters", filter.to_json ]
```
That's a nice touch :)

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+  end
+  
+  def create_export_folder
+    Dir.mkdir("#{Rails.root}/export") unless File.exist?("#{Rails.root}/export")
```
now csv_export and the job know about the export folder -> only csv_export should know -> make it a constant or make it `File.dirname(csv_export.full_filename)`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+    end_date = Date.today
+    orig_dates = [ start_date, end_date]
+
+    if params.key?(:start_date)
```
`if start_date = params[:start_date].presence`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+
+    if params.key?(:start_date)
+      begin
+        start_date = date_from_params params[:start_date]
```
I'd recommend to fail fast / not hide errors ... blow up ... if something goes wrong here, that means the UI is buggy ..

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+      end
+    end
+
+    unless start_date == orig_dates[0] && end_date == orig_dates[1]
```
might be simpler to not assign them at the start and then do 
```Ruby
if start_date || end_date
  start_date ||= Date.new
  end_date ||= Date.today
  filter['deploys.created_at'] = (start_date..end_date)
end
```

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+      if params[:production] == 'Yes'
+        filter['stages.production'] = true
+      elsif params[:production] == 'No'
+        filter['stages.production'] = false
```
`else raise "Unknown value for production #{params[:production]}"`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+    if params.key?(:status)
+      status = []
+      params[:status].each do |s|
+        if ['succeeded', 'running', 'failed', 'errored', 'cancelling', 'cancelled'].include?(s)
```
blow up on unknown ...

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+        if ['succeeded', 'running', 'failed', 'errored', 'cancelling', 'cancelled'].include?(s)
+          status << s
+        end
+        if status.length > 0 and status.length < 6
```
the 6 is a magic number ... can be something like `Deploy::STATUS.size`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,154 @@
+      end
+    end
+
+    if params.key?(:project)
```
`if project_id = params[:project].presence` ... and then no more need for .to_i

grosser commented on app/mailers/csv_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+  def created_email(csv_export)
+    address = csv_export.email
+    subject = "Samson CSV Export Completed"
+    url = csv_url( id: csv_export.id, format: 'csv')
```
nitpick: no leading whitespace

grosser replied:
also: `csv_url(csv_export, format: :csv)`

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,57 @@
+    "#{content}_#{formatted_create_at}.csv"
+  end
+
+  def full_filename
```
might be nice to use different names like 'download_name' and `path` ... I was expecting full_filename to be `/export/#{filename}` ...

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,57 @@
+    "#{Rails.root}/export/#{id}"
+  end
+
+  def pending!
```
use a `status!` method to get rid of all these setters ?

grosser replied:
(as the public interface)

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,57 @@
+  
+  private
+
+  def formatted_create_at
```
used once ... does not need to be an extra method ...

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,57 @@
+class CsvExport < ActiveRecord::Base
+  belongs_to :user
+
```
list of possibly status values as constant and then validate that status is in that list

grosser replied:
use `serialize :filters, JSON` to avoid having to do any work yourself :)

grosser commented on app/models/deploy.rb:
```diff
@@ -122,7 +122,7 @@ def self.expired
 
   def csv_buddy
-    if not (stage.deploy_requires_approval?)
+    if not stage.deploy_requires_approval?
```
`unless`

grosser commented on app/views/csvs/index.html.erb:
```diff
@@ -0,0 +1,37 @@
+            <% if csv_export.status? :ready %>
+              <td><%= link_to "Download", csv_path(csv_export, format: 'csv') %></td>
+            <% else %>
+              <td><%= link_to "Status", csv_path(csv_export) %></td>
```
`link_to "Details"` ?

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,42 @@
+      </div>
+      <div>
+        <%= label_tag :end_date, "End Date:" %>
+        <%=	select_date Date.today, prefix: :end_date %>
```
nitpick: leading whitespace

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,42 @@
+      <div>Configure your report with these parameters</div>
+      <div>
+        <%= label_tag :start_date, "Start Date:" %>
+        <%= select_date Date.today.prev_year, prefix: :start_date %>
```
all these could be `filters[start_date]` etc and then we would send less 1-off things into the csv export columns

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,42 @@
+      <div>
+        <%= label_tag :project, "Project" %>
+        <% projects = [['all',0]] %>
+        <% projects += Project.all.map { |p| [p.name, p.id]} %>
```
`.all` load the full records, but we only need to columns -> `.pluck(:name, :id)` ... and then then `.map` can go away too

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,42 @@
+        <% projects += Project.all.map { |p| [p.name, p.id]} %>
+        <%= select_tag(:project, options_for_select(projects, 0)) %>
+      </div>
+      <%= hidden_field_tag :content, 'deploys' %>
```
it's always deploys ?

rikeoka replied:
For the deploys report it will.  I'm going to have to get a report on stages to see what stages have buddy deploys set for reporting as well.  That's going to be the second report that hits the controller, but it likely be a realtime report since it's not processing thousands of records.

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,42 @@
+    <% end %>
+  </div>
+  <% if current_user.is_admin? %>
+      <hr />
```
nitpick: 2-space indent

grosser commented on app/views/csvs/show.html.erb:
```diff
@@ -0,0 +1,16 @@
+
+<section class="clearfix">
+  <div>
+    <% if @csv_export.status? :deleted? %>
```
should be `:deleted` ?

grosser commented on app/views/csvs/show.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <% if @csv_export.status? :deleted? %>
+      <p>Your report has been deleted.  <%= link_to "Request a new report.", new_csv_path %></p>
+    <% elsif @csv_export.status? :ready %>
+      <p><%= link_to "Your report is ready for download.", csv_path(@csv_export, format: 'csv') %></p>
```
no need for chit-chat :D 
`link_to "Download", ...`

grosser commented on app/views/csvs/show.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <% elsif @csv_export.status? :ready %>
+      <p><%= link_to "Your report is ready for download.", csv_path(@csv_export, format: 'csv') %></p>
+    <% elsif @csv_export.status? :failed %>
+      <p>Your report has failed.  <%= link_to "Request a new report.", new_csv_path %></p>
```
can we store the error ... and dump that here ... or say 'please contact and administrator to check the logs'

grosser replied:
dumping the error might not be a good idea for security reasons ... for now just say 'please contact and administrator to check the logs export #{export.id}'

grosser commented on app/views/deploys/recent.html.erb:
```diff
@@ -29,7 +29,7 @@
   <section class="timeline-content clearfix">
     <div class="deploy-csv">
-      <%= link_to "Download as CSV", recent_deploys_path(format: "csv") %>
+    	<%= link_to "Request a deploys report", new_csv_path %>
```
2-space indent

rikeoka commented:
Updated new deploys report UI
<img width="437" alt="screen shot 2016-03-22 at 6 04 45 pm" src="https://cloud.githubusercontent.com/assets/11447948/13972785/14912a36-f059-11e5-8c2d-fb3432897a46.png">
<img width="301" alt="screen shot 2016-03-22 at 6 10 57 pm" src="https://cloud.githubusercontent.com/assets/11447948/13972835/79fd488c-f059-11e5-8427-fe4f5cd5bcb2.png">

grosser commented on app/assets/javascripts/cvss.js:
```diff
@@ -0,0 +1,13 @@
+(function() {
```
this should do ... ideally use a class to make this `.datepicker` or check if someone already did that
```js
jQuery(function() {
   $("#datepicker1, #datepicker2").datepicker();
})
```

sandlerr replied:
this filename is incorrect

rikeoka replied:
Already declared in changelog.js so relying on that for the date picker and updated the id to work with that one.

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+    respond_to do |format|
+      format.json { render json: {status: "not found"}.to_json, status: :not_found }
+      format.csv  { render body: "not found", status: :not_found }
+      format.html { render file: "#{Rails.public_path}/404.html", status: :not_found }
```
what happens if you do `raise exception` for the html part ... does it render the normal 404 page ?

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+  def create
+    filters = deploy_filter(params)
+    csv_export = CsvExport.create!(user: current_user, filters: filters.to_json)
+    csv_export.status! :pending
```
the csv_export can take care of this with a before_create callback / by having pending as a default value

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+    csv_export = CsvExport.create!(user: current_user, filters: filters.to_json)
+    csv_export.status! :pending
+    CsvExportJob.perform_later(csv_export.id)
+    redirect_to csv_path(csv_export)
```
`redirect_to csv_export` should do the trick

rikeoka replied:
Was getting NoMethodError: undefined method `csv_export_url'.  Do i need to do something to the routes to make this work?  I think it's because the model is named csv_export, but didn't want to have a model named csv.

grosser replied:
because the routes do not match the models -> csv_exports_controller + csv_export + /csv_exports

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+      begin
+        start_date = Date.strptime(param, "%m/%d/%Y")
+      rescue
+        raise("Invalid start date #{param}.")
```
might as well let the original exception go through ... won't show to the user anywy and should never happen since we use a fancy UI + we get better error details

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+
+    if param = params[:production].presence
+      if param == 'Yes'
```
smells like a case statement

```Ruby
case param
when "Yes" then filter['stages.production'] = true
when "No" then filter['stages.production'] = false
when "Any" then # ignore
else
  raise "Invalid production filter #{param}"
end
```

grosser commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+
+    if param = params[:status].presence
+      if ['succeeded', 'failed'].include?(param.downcase)
```
don't need to downcase if we control the input ?

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+  
+  def create_export_folder(csv_export)
+    Dir.mkdir(File.dirname(csv_export.path_file)) unless File.exist?(File.dirname(csv_export.path_file))
```
that should do the trick:
```Ruby
FileUtils.mkdir_p(File.dirname(csv_export.path_file))
```

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,39 @@
+class CsvExport < ActiveRecord::Base
+  belongs_to :user
+  STATUS_VALUES = ['pending', 'started', 'finished', 'downloaded', 'failed', 'deleted']
```
could add a `validates :status, in: STATUS_VALUES` for safety ...

sandlerr replied:
or you can use [activerecord enum](http://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html)

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,39 @@
+  belongs_to :user
+  STATUS_VALUES = ['pending', 'started', 'finished', 'downloaded', 'failed', 'deleted']
+
+  before_destroy :file_delete
```
nitpick: `delete_file` makes more sense to me ...

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def filters
+    filter = JSON.parse(super)
```
FYI will blow up on nil or empty ... using the `serializes :filters, JSON` might be easier to read

grosser commented on app/models/deploy.rb:
```diff
@@ -127,27 +127,23 @@ def self.expired
-      "Bypassed"
     else
-      buddy.name
+      if buddy.nil? && pending?
```
this looks wrong ... should be `elsif` ... and that won't work work unless ... so
```Ruby
if !stage.deploy_requires_approval?
  ...
elsif ...
```

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,37 @@
+
+<section class="clearfix">
+  <div class="csv-deploy-filters clearfix">
+    <%= form_tag csvs_path do %>
```
this is a create so `form_for CsvExport do |f| ` would be more ideomatic

grosser commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,37 @@
+      </div>
+      <div>
+        <%= label_tag :end_date, "End Date:" %>
+        <%=	date_field_tag :end_date, Date.today.strftime("%m/%d/%Y"), id: "datepicker2" %>
```
whitespace

grosser commented on test/jobs/csv_export_job_test.rb:
```diff
@@ -0,0 +1,122 @@
+      empty_test(filter)
+    end
+
+    it "accurately has no results for date range before deploys" do
```
the `accurately ` is redundant :)

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,112 @@
+describe CsvExport do
+  let(:user) { users(:deployer) }
+
+  describe "CsvExport" do
```
1 `describe CsvExport do` is enough :)

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,112 @@
+      assert_not_empty @csv_export.download_name
+    end
+
+    describe "# email" do
```
nitpick: no space between # and email

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+    end
+
+    describe "file_delete" do
```
`#file_delete`

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+
+      it "deletes file when file_delete called" do
+        assert File.exists?(@filename), "File not created in setup"
```
can move into the `setup`

grosser commented:
Looks good to me :)

@sandlerr want to do a quick check to see what I missed ?

sandlerr commented on app/controllers/csvs_controller.rb:
```diff
@@ -0,0 +1,107 @@
+        send_file @csv_export.path_file, type: :csv, filename: @csv_export.download_name
+        @csv_export.status! :downloaded
+        Rails.logger.info("#{current_user.name_and_email} just downloaded #{@csv_export.download_name})")
+      rescue
```
can you rescue a specific exception class here?

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+      csv << ["Deploy Number", "Project Name", "Deploy Sumary", "Deploy Commit", "Deploy Status", "Deploy Updated",
+        "Deploy Created", "Deployer Name", "Deployer Email", "Buddy Name", "Buddy Email",
+        "Production Flag", "No code deployed" ]
+      Deploy.uncached do
```
why was this needed?

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+ 
+  def perform(csv_export_id)
+    ActiveRecord::Base.connection_pool.with_connection do
+      csv_export = CsvExport.find(csv_export_id)
```
I'm pretty sure ActiveJob can actually do this for you - can't you just pass in the model rather than its id?

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+    notify_of_creation(csv_export)
+  rescue
+    csv_export.status! :failed
+    Rails.logger.info("Export #{csv_export.download_name} failed")
```
would be good to log the actual error as well as report it to airbrake

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+    # TODO delete files with no record associations
+    end_date = Time.now - Rails.application.config.samson.export_job.downloaded_age
+    timeout_date = Time.now - Rails.application.config.samson.export_job.max_age
+    @csv_exports = CsvExport.where("(status = 'downloaded' AND updated_at <= :end_date) OR updated_at <= :timeout_date",
```
this could be a scope on the model

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,72 @@
+      {end_date: end_date, timeout_date: timeout_date})
+    @csv_exports.each do |csv_export|
+      csv_export.destroy
+      Rails.logger.info("Downloaded file #{csv_export.download_name} deleted")
```
this will always log but the file is only deleted if it exists (and the destroy succeeds, which we aren't checking here)

rikeoka replied:
Moving the logging info into the delete_file function.  That way it'll state if it succeeds or fails.

sandlerr commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def email
+    user.try :email
```
can replace this method by something like `delegate :email, to: :user, allow_nil: true`

sandlerr commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+  
+  def download_name
+    "deploys_#{created_at.strftime "%Y%m%d_%H%M%S"}.csv"
```
could also use epoch time `created_at.to_s(:number)` to avoid the edge case where two are created in one second

sandlerr commented on app/models/deploy.rb:
```diff
@@ -127,27 +127,23 @@ def self.expired
-          csv << [deploy.id, deploy.project.name, deploy.summary, deploy.updated_at, deploy.start_time, deploy.job.user.name, deploy.csv_buddy, deploy.stage.production]
-        end
-    end
+  def buddy_email
```
can `delegate` this too

rikeoka replied:
We only wanted to include this if buddy approval is required?  Can't seem to find a way to make that work.

sandlerr replied:
won't it just be `nil` if there is no buddy?

rikeoka replied:
I did some refactoring of csv_buddy and buddy_email to have consistent results, with only 3 outcomes buddy is user results in bypassed, or the value or nil.

sandlerr commented on app/views/csvs/new.html.erb:
```diff
@@ -0,0 +1,37 @@
+      <div><strong>Deploy reports</strong></div>
+      <div>
+        <%= label_tag :start_date, "Start Date:" %>
+        <%= date_field_tag :start_date, Date.today.prev_year.strftime("%m/%d/%Y"), id: "datepicker1" %>
```
fyi this date format is not standard across all locales :)

grosser replied:
... might be better to go with %Y-%m-%d ...

On Wed, Mar 23, 2016 at 5:03 PM, Roman Sandler <notifications@github.com>
wrote:

> In app/views/csvs/new.html.erb
> <https://github.com/zendesk/samson/pull/766#discussion_r57257656>:
>
> > @@ -0,0 +1,37 @@
> > +<h1>Request a CSV Report</h1>
> > +
> > +<section class="clearfix">
> > +  <div class="csv-deploy-filters clearfix">
> > +    <%= form_tag csvs_path do %>
> > +      <div><strong>Deploy reports</strong></div>
> > +      <div>
> > +        <%= label_tag :start_date, "Start Date:" %>
> > +        <%= date_field_tag :start_date, Date.today.prev_year.strftime("%m/%d/%Y"), id: "datepicker1" %>
>
> fyi this date format is not standard across all locales :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zendesk/samson/pull/766/files/fbb4d120d26ded64e1ff88a3485de7c2957cd7fd#r57257656>
>

sandlerr commented on app/views/csvs/show.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <% elsif @csv_export.status? :ready %>
+      <p><%= link_to "Download", csv_path(@csv_export, format: 'csv') %></p>
+    <% elsif @csv_export.status? :failed %>
+      <p>Your report has failed.  Please contact an administrator to check the logs export <%=@csv_export.id%>.</p>
```
can you add whitespaces around `@csv_export.id`?

sandlerr commented on config/initializers/sucker_punch.rb:
```diff
@@ -0,0 +1 @@
+require 'sucker_punch/async_syntax'
```
does ActiveJob require this?

rikeoka replied:
This is needed until Rails 5.0, when the queue adapter is changed to async_perform from new.async.perform.  Could add a comment in the initializer that it can be removed when Rails is 5.0 or newer.

sandlerr replied:
sure, or you could raise if the rails version is >=5 :)

rikeoka replied:
Added in 7840b4b

sandlerr commented on app/controllers/csvs_controller.rb:
```diff
@@ -41,7 +40,7 @@ def download
         Rails.logger.info("#{current_user.name_and_email} just downloaded #{@csv_export.download_name})")
-      rescue
+      rescue ActionController::MissingFile
         @csv_export.status! :deleted
```
actually, won't this codepath be hit if a request comes in the small window of time if a file is `pending` and hasn't been created yet by the worker?

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -31,9 +24,11 @@ def export_task(csv_export)
     deploy_csv_export(csv_export)
     notify_of_creation(csv_export)
-  rescue
+  rescue Exception => e
```
rescuing `Exception` is generally not a good idea. Take a look at http://ruby-doc.org/core-2.2.0/Exception.html to see all the things this might be swallowing - among other things it eats OOM errors and signals

sandlerr commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -0,0 +1,102 @@
+    # sanitizes parameters and generates a filter string for use with the Deploy.joins(:stage, :jobs)
+    filter = {}
+
+    if param = params[:start_date].presence
```
nitpick: I would prefer if these weren't all called `param`, maybe use more descriptive names like `start_date`, `project_id`, etc

sandlerr commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def filters
+    filter = super
+    if filter.nil? || filter.class == "".class
```
how can it be a string?

sandlerr replied:
also you could use `filter.is_a?(String)`

rikeoka replied:
Was having a large amount of parsing errors in the rake test, because of the table required filters to be not null.  Seems to work now without it.

sandlerr commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,64 @@
+    filter = super
+    if filter.nil? || filter.class == "".class
+      filter = {}
+    elsif filter.fetch('deploys.created_at', nil).class == "".class
```
falling back to nil is the same as `filter['deploy.created_at']` isn't it?

sandlerr commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,65 @@
+    csv_export.status! :started
+    deploy_csv_export(csv_export)
+    notify_of_creation(csv_export)
+  rescue NoMethodError, IOError, ActiveRecord::ActiveRecordError => e
```
`NoMethodError`? Because of calling a method on `nil`? I can't see where it would be raised.

As an alternative, would you prefer to `rescue` generically, do your cleanup and then re-raise?

rikeoka replied:
I think I eliminated all cases where calling a method on nil would occur.  The report has to pull data from several linked models.  I don't think it would ever come up since I replaced the known cases with try, but it may still occur.  

If it gets eliminated and not caught the job would simply fail with no cleanup, and the user won't get the report because it would never reach a finished state.  The user may wonder why the report is taking so long, however.  

The no method error was coming up because of deleted projects, stages or users.  I think now with the dependent delete/destroys it wouldn't appear for newer records, but with some of the older historical records it may come up.

sandlerr replied:
I wonder if we should explicitly check for deleted records or include them using `Project.with_deleted { ... }`. This code was a bit confusing to me because in most cases `NoMethodError` is a bug

rikeoka replied:
Going to remove the NoMethodError.  The issue was with the behavior with the deleted projects was fixed back in August 2014 29ba57b.  The join with stages actually prevents any deleted projects from causing issues.  The only other potential source of errors would be from the deployer email, which is fixed by using try(:email).

sandlerr commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def filters
+    filter = super || {}
+    if filter['deploys.created_at'].class == "".class
```
maybe `filter['deploys.created_at'].is_a?(String)` ?

sandlerr commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -0,0 +1,102 @@
+    end
+  end
+
+  def deploy_filter(params)
```
you shouldn't pass in params, it will already be available in private controller methods just as it is in an action

sandlerr commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -0,0 +1,102 @@
+      end
+    end
+
+    if project = params[:project].presence
```
would `params[:project].try(:to_i)` work here?

rikeoka commented:
PR#759 went into prod.  Still unable to get report using the search API, due to large volume of deploys in production.

grosser commented on .env.example:
```diff
@@ -50,6 +50,19 @@ GITHUB_TOKEN={fill me in}
+## EXPORT_JOB_DOWNLOADED_AGE determines how long a csv export job and file should
+## persist after it is downloaded before cleanup, default is 12.hours.
+##
+## EXPORT_JOB_MAX_AGE determines how long a csv export job and file should persist
```
is there a usecase where you request a download and then come back a month later to download it ?
can we delete all downloads after 24 hours for example ?

rikeoka replied:
I don't see the need anymore because of customizable filters on the report.  I think 1 day is good for the default cleanup.  If 1 day is too long or short, changing the env value could be done without a code change now.

grosser commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -0,0 +1,102 @@
+
+    if production = params[:production].presence
+      case production
+        when 'Yes' then filter['stages.production'] = true
```
outdent once, `case` and `when` should be on the same line

grosser commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -0,0 +1,102 @@
+      if ['succeeded', 'failed'].include?(status)
+        filter['jobs.status'] = status
+      elsif status != "all"
+        raise "Invalid status filter #{status}"
```
❤️  all this fast failing :)

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+  end
+  
+  def export_task(csv_export)
```
maybe `def generate_csv`

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,65 @@
+  
+  def notify_of_creation(csv_export)
+    CsvMailer.created_email(csv_export).deliver_now if csv_export.email.present?
+    csv_export.status! :finished
```
these 2 lines don't have anything to do with 'notifying' ... maybe better in the method above

rikeoka replied:
Just moved all 3 lines above, no point in a separate function since it's used only once.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,65 @@
+    Rails.logger.info("Export #{csv_export.download_name} completed")
+  end
+  
+  def deploy_csv_export(csv_export)
```
deploy -> write/dump ?

grosser replied:
ah ... this is a dump of the deploys ... not 'deploying' ...

rikeoka replied:
Didn't seem to have anything to fix

grosser commented on app/mailers/csv_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+class CsvMailer < ApplicationMailer
+  def created_email(csv_export)
```
all methods are emails -> no need for the _email suffix (like naming controller methods `def show_action`)

grosser commented on app/mailers/csv_mailer.rb:
```diff
@@ -0,0 +1,9 @@
+    address = csv_export.email
+    subject = "Samson CSV Export Completed"
+    url = csv_export_url(csv_export, format: 'csv')
+    body = "Download your CSV file at #{url}"
```
`The csv export you requested on #{created_at} has finished!`

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+class CsvExport < ActiveRecord::Base
+  belongs_to :user
+  serialize :filters, JSON
+  STATUS_VALUES = [nil, 'pending', 'started', 'finished', 'downloaded', 'failed', 'deleted']
```
nil is a valid status ?

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+
+
+  scope :old, lambda {
+    end_date = Time.now - Rails.application.config.samson.export_job.downloaded_age
```
`.seconds.ago`

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+    end_date = Time.now - Rails.application.config.samson.export_job.downloaded_age
+    timeout_date = Time.now - Rails.application.config.samson.export_job.max_age
+    where("(status = 'downloaded' AND updated_at <= :end_date) OR updated_at <= :timeout_date",
+    {end_date: end_date, timeout_date: timeout_date})
```
2-space indent and no need for `{}`

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def filters
+    filter = super || {}
+    if filter['deploys.created_at'].is_a?(String)
```
the controller should handle that / do not let bad data get stored in the first place

rikeoka replied:
Worked out the original issue I was having, apparently the tests was failing without the check for the String it fails the test because of a state issue.  When loaded from the DB, deploys.created_at is a string, once this function runs the state of object makes this a range.

refactoring into a differently named method it now works correctly without state issues.

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def delete_file
+    if File.exist?(path_file)
+      File.delete(path_file)
```
that should blow up when it does not work ... don't think we need the extra check after ...

grosser commented on db/migrate/20160218161746_add_csv_export.rb:
```diff
@@ -0,0 +1,10 @@
+      t.integer :user_id,    null: false
+      t.timestamps           null: false
+      t.text :filters
+      t.string :status
```
set default to 'pending' here

grosser commented on db/migrate/20160218161746_add_csv_export.rb:
```diff
@@ -0,0 +1,10 @@
+    create_table :csv_exports do |t|
+      t.integer :user_id,    null: false
+      t.timestamps           null: false
+      t.text :filters
```
set default to {} here

rikeoka replied:
Had to change to string to make it work with mysql.  I don't see a case where the filter string will grow to > 255 characters now that the filter is controlled.

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,62 @@
+
+  private
+
+  def set_defaults
```
can use migration for that -> always consistent and less code

grosser commented on test/mailers/csv_mailer_test.rb:
```diff
@@ -0,0 +1,17 @@
+require_relative '../test_helper'
+
+describe CsvMailer do
+  setup do
```
prefer `before` ... kind of does the same, but fits the spec style better

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,136 @@
+  let(:user) { users(:deployer) }
+  setup { @csv_export = CsvExport.create(user: user) }
+
+  it "sets pending upon creation" do
```
`describe "#status?" do`

grosser replied:
or `describe "status methods"` since this is all mixed up

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,136 @@
+      csv_export.status.must_equal status
+      csv_export.filters.must_equal filters
+    end
+
```
extra newline

grosser commented:
this is starting to look really nice!

rikeoka commented:
@grosser app/jobs/csv_export_job_test.rb seems to be failing SingleCov.covered! even with file: 'app/jobs/csv_export_job.rb' option added.  It also seems to be guessing lib/jobs/csv_export_job.rb which may be why it's failing

grosser commented:
bundle update single_cov should take care of it :)

rikeoka commented:
It worked:)

grosser commented on app/models/csv_export.rb:
```diff
@@ -0,0 +1,47 @@
+  def filters
+    filter = super.clone
+    if filter['deploys.created_at']
+      dates = filter['deploys.created_at'].scan(/-?\d+-\d+-\d+/)
```
should the controller handle this instead ... it does all the filtering, right ?

rikeoka replied:
l3:     serialize :filters, JSON 
can't handle the date range properly.  This gets parsed into a string.  This is simply reformatting (date..date) into the correct date range.  Not sure if I can pass something to the JSON parser so it handle this correctly.

grosser replied:
ahh that makes sense ... at least this way the data coming in/out of csv_export is the same :)

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -0,0 +1,53 @@
+      <div class="form-group">
+        <%= label_tag :production, "Production:", class: "col-lg-2 control-label" %>
+        <div class="col-lg-2">
+          <%= select_tag :production, options_for_select(['Any', 'Yes', 'No'], 'Any'), class: "form-control" %>
```
FYI I usually opt for blank to mean any/all so the UI stays cleaner ... but not important ...

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -0,0 +1,53 @@
+      <div class="form-group">
+        <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
+        <% projects = [['All',0]] %>
+        <% projects += Project.all.pluck(:name, :id) %>
```
- No need for the `.all`
 - FYI prefer `.concat` to `+=`, it avoid creating extra objects

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -0,0 +1,53 @@
+      </div>
+      <div class="form-group">
+        <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
+        <% projects = [['All',0]] %>
```
nitpick: space after comma

grosser commented on app/views/shared/_header.html.erb:
```diff
@@ -51,10 +51,10 @@
-          <a href="#" class="dropdown-toggle" data-toggle="dropdown">Admin <b class="caret"></b></a>
-          <ul class="dropdown-menu">
+      <li class="dropdown <%= 'active' if request.path.starts_with?('/admin') %>">
+        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Admin <b class="caret"></b></a>
```
normal users should not see the admin dropdown ... they might get scared and think it's a bug :(

grosser replied:
maybe best as a separate tab ... or name it `More` ... or `Manage`

rikeoka replied:
Separate tab had a comment it shouldn't be a top level menu item, so going with Manage

grosser commented on config/initializers/sucker_punch.rb:
```diff
@@ -0,0 +1,2 @@
+require 'sucker_punch/async_syntax'
+raise "Require in config/initializers/sucker_punch.rb is no longer needed" if Rails::VERSION::STRING.split('.')[0].to_i >= 5
```
`Rails::VERSION::MAJOR >= 5`

grosser commented on db/migrate/20160218161746_add_csv_export.rb:
```diff
@@ -0,0 +1,10 @@
+    create_table :csv_exports do |t|
+      t.integer :user_id,    null: false
+      t.timestamps           null: false
+      t.string :filters, default: "{}"
```
`null: false` for status and filters just in case ...

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,259 @@
+
+  def show_expected_html(state)
+    case state.to_s
+      when 'deleted'
```
nitpick: case and when should align, then the else will align with the end :)

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,259 @@
+  end
+
+  as_a_deployer do
+    describe "#Index" do
```
nitpick: lowercase

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,259 @@
+
+  describe "permissions test" do
+    as_a_admin do
+      describe "a GET to new" do
```
all others are `describe "#XXX` ... this one should too :)

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,259 @@
+              it "changes state to deleted if ready" do
+                assert export.reload.status? :deleted
+              end
+            end
```
could do the opposite test for else: `it 'does not change state'`

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,259 @@
+  end
+
+  def cleanup_files
+    CsvExport.all.each { |csv_file| csv_file.destroy! }
```
FYI `CsvExport.find_each(&:destroy!)`

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,149 @@
+    it "sets defaults" do
+      csv_export = CsvExport.create(user: user)
+      csv_export.status.must_equal "pending"
+      csv_export.filters.must_equal Hash.new
```
FYI alternatively `.must_equal({})`

rikeoka replied:
Was trying this without the () originally.

grosser commented on app/jobs/csv_export_job.rb:
```diff
@@ -0,0 +1,60 @@
+  private
+
+  def cleanup_downloaded
+    CsvExport.old.each do |csv_export|
```
`.old.find_each(&:destroy)`

rikeoka replied:
updated

grosser replied:
Ideally comment / update after you pushed :)

rikeoka replied:
Updated with an amended commit before I merged the master onto the branch, but undid the amend to make the fix clearer:)

rikeoka commented:
Updated with the latest master changes and fixed couple of things that broke from that.  

Updated menu for admin:
<img width="179" alt="screen shot 2016-05-11 at 5 53 13 pm" src="https://cloud.githubusercontent.com/assets/11447948/15201009/439cc244-17a1-11e6-819e-aca18918ed28.png">

Updated menu when not an admin:
<img width="158" alt="screen shot 2016-05-11 at 5 54 36 pm" src="https://cloud.githubusercontent.com/assets/11447948/15201030/7784c1ec-17a1-11e6-9da6-70194a702460.png">

grosser commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -0,0 +1,270 @@
+    end
+  end
+
+  as_a_deployer do
```
there are no permission checks, so this would also work with `as_a_viewer`

rikeoka replied:
Made the as_a_deployer, as_a_viewer to better detect future permission changes.

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,149 @@
+  let(:user) { users(:deployer) }
+  before { @csv_export = CsvExport.create(user: user) }
+
+  describe "scopes" do
```
`describe ".old"`

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,149 @@
+    end
+  end
+
+  describe "delete_file" do
```
`describe "#delete_file"`

grosser commented on test/models/csv_export_test.rb:
```diff
@@ -0,0 +1,149 @@
+  end
+
+  it "sets filename if not assigned" do
+    assert_not_empty @csv_export.download_name
```
part of defaults test ?

rikeoka replied:
Shouldn't be it sets since it doesn't change the record at all.  Updated with describe #download_name and it includes the created_date.

grosser commented:
👍 🎉

grosser commented:
👍  🚢
